File: 프로그래머스/2/49994. 방문 길이/방문 길이.py
def solution(dirs):
    answer = 0
    start = (0,0)
    UDRL = {"U" : [0,1],"D" : [0,-1],"R" : [1,0],"L" : [-1,0]}
    countSet = set()
    for i in dirs:
#         키에서 벨류갑 봅아서 쓰는법 ->  UDRL[i][0] 일케하면됨
  
        dx, dy = start[0]+ UDRL[i][0], start[1]+UDRL[i][1]
   
        if dx < -5 or dx > 5 or dy < -5 or dy > 5:
            continue
        end = (dx, dy)
#         [기존 좌표, 다음 좌표], [다음좌표, 기존 좌표] set에 넣기
#         list1 = []
#         list1.append(start)
#         list1.append(end)
# #         set에 리스트는 못 넣는다
#         countSet.add(tuple(list1))
#         list1 = []
#         list1.append(end)
#         list1.append(start)
#         countSet.add(tuple(list1))
        countSet.add((start[0],start[1],dx,dy))
        countSet.add((dx,dy,start[0],start[1]))
        start = end       
        # print(countSet)

        
    answer = len(countSet) // 2
    return answer

File: 프로그래머스/2/49994. 방문 길이/방문 길이.cpp
#include <string>
#include <vector>
#include <set>
#include <iostream>
using namespace std;

int solution(string dirs) {
    int answer = 0;
    set<vector<int>> st;
    int x = 0;
    int y = 0;
    for(char dir : dirs){
        if(dir == 'U'){
            int ny = y+1;
            if(ny <=5){
                st.insert({x,y,ny,x});
                st.insert({x,ny,y,x});
                y=ny;
            } else{
                continue;
            }
        } else if(dir =='D'){
            int ny = y-1;
            if(-5<=ny){
                st.insert({x,y,ny,x});
                st.insert({x,ny,y,x});
                y = ny;
            }else{
                continue;
            }
        } else if(dir == 'R'){
            int nx = x+1;
            if(nx <= 5){
                st.insert({nx,y,y,x});
                st.insert({x,y,y,nx});
                x = nx;
            }else{
                continue;
            }
        } else{
            int nx = x-1;
            if(-5<=nx){
                st.insert({nx,y,y,x});
                st.insert({x,y,y,nx});
                x = nx;
            }else{
                continue;
            } 
        }    
    }

    answer = st.size()/2;
    return answer;
}

// {0,0,-1,0} {-1,0,0,0}
// 정답 set<vector<int>> 로넣고 양방향넣자 ,-5<=x<=5 -5<=y<=5
// 1. for문 dirs길이만큼
// 2. UDRL 분기처리  
// 3. 만약 -5<=nx<=5 -5<=ny<=5이라면
//    set에 {출발점, 도착점}{도착점,출발점}
//      x=nx y=ny

// 4. set길이 반환/2